### PR TITLE
MOTECH-2884 Extend Date Manipulation Functions in Tasks

### PR DESCRIPTION
**Pull request by piopawel**
Added 4 new date manipulation functions: Plus month, Minus month, Start of month, End of month. Added a description of them to "Help:modifications" window

### CODE REVIEW

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by sebbrudzinski** [comment]
test this please

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/util/KeyEvaluator.java [hunk 1]
```diff
@@ -266,6 +285,32 @@ private String splitManipulation(String value, String manipulation) {
         return value.split(regex)[idx];
     }
 
+    private String beginningOfMonthManipulation(String value) {
+        DateTime dateTime = new DateTime(value);
+
+        return dateTime.dayOfMonth().withMinimumValue().toString();
```
Comment (sebbrudzinski) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/util/KeyEvaluator.java]: I believe we should set the time to midnight as well.

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/util/KeyEvaluator.java [hunk 2]
```diff
@@ -266,6 +285,32 @@ private String splitManipulation(String value, String manipulation) {
...
+    }
+
+    private String endOfMonthManipulation(String value) {
+        DateTime dateTime = new DateTime(value);
+
+        return dateTime.dayOfMonth().withMaximumValue().toString();
```
Comment (sebbrudzinski) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/util/KeyEvaluator.java]: Let's set the time to 23:59:59

File: modules/tasks/tasks/src/main/resources/webapp/partials/help/manipulation.html [hunk 3]
```diff
@@ -87,9 +102,14 @@ <h4 class="modal-title" id="helpModalLabel">{{msg('task.helpManipulation')}}</h4
...
     <tr>
+        <td>minusMonths(1)</td>
+        <td>{{msg('task.date')}}</td>
+        <td>Wed, July 10, '96' -> Mon, June 10, '96'</td>
+    </tr>
+    <tr>
         <td>minusDays(3)</td>
         <td>{{msg('task.date')}}</td>
-        <td>Wed, July 10, '96' -> Wed, July 7, '96'</td>
+        <td>Wed, July 10, '96' -> Sun, July 7, '96'</td>
```
Comment (sebbrudzinski) [inline modules/tasks/tasks/src/main/resources/webapp/partials/help/manipulation.html]: Thanks for fixing those two! :)

File: modules/tasks/tasks/src/main/resources/webapp/partials/widgets/date-manipulation.html [hunk 4]
```diff
@@ -1,10 +1,28 @@
...
         <span class="glyphicon glyphicon-sort" title="{{msg('task.infoChangeOrder')}}" style="display: none;"></span>
     </li>
+    <!--<li unselectable="on" class="padding-botton6" type="beginningOfMonth">
+        <span unselectable="on" class="pointer" setmanipulation="beginningOfMonth">{{msg('task.beginningOfMonth')}}</span>
+        <span class="glyphicon glyphicon-sort" title="{{msg('task.infoChangeOrder')}}" style="display: none;"></span>
+    </li>
+    <li unselectable="on" class="padding-botton6" type="endOfMonth">
+        <span unselectable="on" class="pointer" setmanipulation="dateTime">{{msg('task.endOfMonth')}}</span>
+        <span class="glyphicon glyphicon-sort" title="{{msg('task.infoChangeOrder')}}" style="display: none;"></span>
+    </li>-->
```
Comment (sebbrudzinski) [inline modules/tasks/tasks/src/main/resources/webapp/partials/widgets/date-manipulation.html]: Do we need this commented out code?

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/util/KeyEvaluator.java [hunk 5]
```diff
@@ -266,6 +285,32 @@ private String splitManipulation(String value, String manipulation) {
...
+
+        return dateTime.plusMonths(Integer.parseInt(pattern)).toString();
+    }
+
+    private String minusMonthsManipulation(String value, String manipulation) {
+        String pattern = manipulation.substring(MINUS_MONTHS_PATTERN_BEGIN_INDEX, manipulation.length() - 1);
+        DateTime dateTime = new DateTime(value);
+
+        return dateTime.minusMonths(Integer.parseInt(pattern)).toString();
+    }
```
Comment (sebbrudzinski) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/util/KeyEvaluator.java]: Please add tests for the new manipulatons.

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
test this please

**Comment by piopawel** [comment]
update branch

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
finally...

**Pull request merged**